### DISCUSSION TITLE
SQLite Instead of snippets.json

### DISCUSSION THREAD

**Discussion by cwarden** [maaslalani/nap#32] [Ideas]
After the migration of snippets to folders with #31, I realized that I have a bunch of orphaned snippets.  This is presumably from running multiple instances of nap concurrently.

Replacing snippets.json with a SQLite database with allow multiple nap processes to update the snippets "database".

**Comment by maaslalani** [comment]
I like this idea, but in all honesty even the snippets.json might be overkill for this project. We could simply stuff the metadata of the snippet, i.e. notes in the file or have a separate hidden file for each snippet with the same name to keep metadata that way we don't need to manage a database.

What are your thoughts on that?

**Reply by cwarden** [reply]
I like it.  I'm not sure any (non-filesystem) metadata is needed.

**Reply by jolheiser** [reply]
I also like this idea, as I've already noticed that parts of the JSON are actually unused/irrelevant to how the code processes things.

I was just thinking of this earlier as well and considered the hidden file approach, however a con I came up with is that there's not as easy of a way to gather all snippets other than scanning. 

That being said, the code in #33 could easily be refactored to instead just be the source of truth for gathering snippets at startup, and it could also easily check each file for an accompanying meta file `foo/bar.baz.yaml` or similar.

One other con is that "external editors" as mentioned in #24 either need to also provide a hidden meta file, or `nap` has to be able to 
a. Create the file if it doesn't exist, or
b. Gracefully handle missing meta files
I think I would prefer the latter, just to avoid unnecessary files. Missing meta file would just mean default meta (i.e. no tags, etc.)

**Reply by maaslalani** [reply]
100% agree on handling missing file data. In other words this metadata would be optional, only if people wanted to attach notes or other stuff to it.

**Reply by maaslalani** [reply]
I'm happy to forego the metadata entirely if you all agree it's unneeded. It can always be added later when we actually need to attach metadata. I was just trying to future proof the system in case we added notes / details but I guess these could be added by the user as comments to the file.

**Reply by jolheiser** [reply]
Perhaps an initial refactor can leave out meta. The parts being used can be inferred currently by folder, file(.ext), and modtime.

I do think meta will be great to add in the future though, I can imagine tags being very useful.

EDIT: Technically `nap` currently stores created time, but I think modtime is more helpful in most cases? Otherwise created time could be the first meta addition.

**Reply by xtrasmal** [reply]
How many times do you edit a snippet? I've got a 5 year old snippet base which I hardly ever alter. 
So I was thinking in just storing a second or third file of that same snippet. 
You can have a background process creating a cache to read from. Containing the reference pointer to the last created recent snippets. The cache could have meta data as well. 
Nap interface can be used to read through all versions of that particular snippet.

**Comment by xtrasmal** [comment]
SQLite is portable, but I'd rather have the snippets saved separately  and individually in some text format.
Nap should be a tool to manage those files, but without nap, I should be able to use whatever crazy tool or script out there to read/transform/edit etc my snippets. Loosely coupled.
So I do like the idea of not storing anything in json. Rather many loose files.
